pkg/template/event_test.go

```diff
@@ -0,0 +1,309 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package template
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"testing"
+
+	pipelinev1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+func Test_EventPathVarRegex(t *testing.T) {
+	tests := []struct {
+		eventPathVar string
+		wantValid    bool
+	}{
+		// Valid
+		{wantValid: true, eventPathVar: "$(event)"},
+		{wantValid: true, eventPathVar: "$(event.a-b)"},
+		{wantValid: true, eventPathVar: "$(event.a1)"},
+		{wantValid: true, eventPathVar: "$(event.a.b)"},
+		{wantValid: true, eventPathVar: "$(event.a.b.c)"},
+		// Invalid
+		{wantValid: false, eventPathVar: "$event"},
+		{wantValid: false, eventPathVar: "$[event]"},
+		{wantValid: false, eventPathVar: "${event}"},
+		{wantValid: false, eventPathVar: "$(event..)"},
+		{wantValid: false, eventPathVar: "$(event.$a)"},
+		{wantValid: false, eventPathVar: "event.a"},
+		{wantValid: false, eventPathVar: "event"},
+		{wantValid: false, eventPathVar: "${{event}"},
+		{wantValid: false, eventPathVar: "${event"},
+	}
+	for _, tt := range tests {
+		name := fmt.Sprintf("valid: %s", tt.eventPathVar)
+		if !tt.wantValid {
+			name = fmt.Sprintf("invalid: %s", tt.eventPathVar)
+		}
+		t.Run(name, func(t *testing.T) {
+			if valid := eventPathVarRegex.MatchString(tt.eventPathVar); valid != tt.wantValid {
+				t.Errorf("eventPathVarRegex.MatchString() = %t, want = %t", valid, tt.wantValid)
+			}
+		})
+	}
+}
+
+func Test_GetEventPathFromVar(t *testing.T) {
+	tests := []struct {
+		eventPathVar string
+		want         string
+	}{
+		{eventPathVar: "$(event)", want: ""},
+		{eventPathVar: "$(event.a-b)", want: "a-b"},
+		{eventPathVar: "$(event.a1)", want: "a1"},
+		{eventPathVar: "$(event.a.b)", want: "a.b"},
+		{eventPathVar: "$(event.a.b.c)", want: "a.b.c"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.eventPathVar, func(t *testing.T) {
+			if eventPath := getEventPathFromVar(tt.eventPathVar); eventPath != tt.want {
+				t.Errorf("getEventPathFromVar() = %s, want = %s", eventPath, tt.want)
+			}
+		})
+	}
+}
+
+var (
+	testEventJSON       = json.RawMessage(`{"one": "onevalue", "two": {"two": "twovalue"}, "three": {"three": {"three": {"three": {"three": "threevalue"}}}}}`)
+	paramNoEventPathVar = pipelinev1.Param{
+		Name:  "paramNoEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar"},
+	}
+	wantParamNoEventPathVar = pipelinev1.Param{
+		Name:  "paramNoEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar"},
+	}
```

Comment:
this makes me wonder if the interface to `applyEventToParam` could be just values in and values out? would be easier to write unit tests for!

i.e. instead of passing a param into `applyEventToParam`, pass a string in, and get a string out

(cut it seems like the name of the param is never relevant to the test case)

Reply:
Hm, that's a good point. We don't actually use the `.Names` anywhere 👍 
However, it will add the annoyance of converting from `[]pipelienv1.Param` -> `[]string` and then back again `[]string` -> `[]pipelinev1.Param`. I'll try it out 🙂